From 6ab131c9d56260ab1bd235d1727efb95f9ff0929 Mon Sep 17 00:00:00 2001
From: Kartatz <105828205+Kartatz@users.noreply.github.com>
Date: Thu, 23 Oct 2025 18:37:34 -0300
Subject: [PATCH] Link libgcobol with libiconv

On Android, we use the iconv routines from an external library rather than from the C library.
---
 gcc/cobol/gcobolspec.cc | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/gcc/cobol/gcobolspec.cc b/gcc/cobol/gcobolspec.cc
index 1f1b4631d..97ffd804a 100644
--- a/gcc/cobol/gcobolspec.cc
+++ b/gcc/cobol/gcobolspec.cc
@@ -69,6 +69,10 @@ int lang_specific_extra_outfiles = 0;
 #define COBOL_LIBRARY "gcobol"
 #endif
 
+#ifndef ICONV_LIBRARY
+#define ICONV_LIBRARY "iconv"
+#endif
+
 #define SPEC_FILE "libgcobol.spec"
 
 /* The original argument list and related info is copied here.  */
@@ -157,6 +161,7 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
 
   bool need_libdl       = (DL_LIBRARY[0] != '\0');
   bool need_libstdc     = (STDCPP_LIBRARY[0] != '\0');
+  bool need_libiconv     = (ICONV_LIBRARY[0] != '\0');
 
   // Separate flags for a couple of static libraries
   bool static_libgcobol  = false;
@@ -256,6 +261,7 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
         need_libgcobol   = false;
         need_libdl       = false;
         need_libstdc     = false;
+        need_libiconv     = false;
         break;
 
       case OPT_static_libgcobol:
@@ -277,6 +283,10 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
           {
           need_libstdc = false;
           }
+        else if(strcmp(decoded_options[i].arg, ICONV_LIBRARY) == 0)
+          {
+          need_libiconv = false;
+          }
         break;
 
       case OPT_o:
@@ -373,6 +383,7 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
     need_libgcobol   = false;
     need_libdl       = false;
     need_libstdc     = false;
+    need_libiconv     = false;
     }
 
   /* Second pass through arglist, transforming arguments as appropriate.  */
@@ -519,6 +530,10 @@ lang_specific_driver (struct cl_decoded_option **in_decoded_options,
     {
     add_arg_lib(STDCPP_LIBRARY, false);
     }
+  if( need_libiconv )
+    {
+    add_arg_lib(ICONV_LIBRARY, false);
+    }
 
   if( prior_main )
     {
-- 
2.50.1

